Neos.RedirectHandler.NeosAdapter/Classes/Service/NodeRedirectionService.php

```diff
@@ -0,0 +1,218 @@
+
+    /**
+     * @Flow\Inject
+     * @var PersistenceManagerInterface
+     */
+    protected $persistenceManager;
+
+    /**
+     * @Flow\Inject
+     * @var ContextFactory
+     */
+    protected $contextFactory;
+
+    /**
+     * @Flow\Inject
+     * @var SystemLoggerInterface
+     */
+    protected $systemLogger;
+
+    /**
+     * @Flow\InjectConfiguration(path="statusCode", package="Neos.RedirectHandler")
+     * @var array
+     */
+    protected $defaultStatusCode;
+
+    /**
+     * Creates a redirection for the node if it is a 'TYPO3.Neos:Document' node and its URI has changed
+     *
+     * @param NodeInterface $node The node that is about to be published
+     * @param Workspace $targetWorkspace
+     * @return void
+     * @throws Exception
+     */
+    public function createRedirectionsForPublishedNode(NodeInterface $node, Workspace $targetWorkspace)
+    {
+        $nodeType = $node->getNodeType();
+        if ($targetWorkspace->getName() !== 'live' || !$nodeType->isOfType('TYPO3.Neos:Document')) {
+            return;
+        }
+
+        $context = $this->contextFactory->create([
+            'workspaceName' => 'live',
+            'invisibleContentShown' => true,
+            'dimensions' => $node->getDimensions()
+        ]);
+        $targetNode = $context->getNodeByIdentifier($node->getIdentifier());
+        if ($targetNode === null) {
+            // The page has been added
+            return;
+        }
+
+        $targetNodeUriPath = $this->buildUriPathForNodeContextPath($targetNode->getContextPath());
+        if ($targetNodeUriPath === null) {
+            throw new Exception('The target URI path of the node could not be resolved', 1451945358);
+        }
+
+        $hosts = $this->getHostPatterns($node->getContext());
+
+        // The page has been removed
+        if ($node->isRemoved()) {
+            $this->flushRoutingCacheForNode($targetNode);
+            $statusCode = (integer)$this->defaultStatusCode['gone'];
+            $this->redirectionStorage->addRedirection($targetNodeUriPath, '', $statusCode, $hosts);
+            return;
+        }
+
+        // compare the "old" node URI to the new one
+        $nodeUriPath = $this->buildUriPathForNodeContextPath($node->getContextPath());
+        // use the same regexp than the ContentContextBar Ember View
+        $nodeUriPath = preg_replace('/@[A-Za-z0-9;&,\-_=]+/', '', $nodeUriPath);
+        if ($nodeUriPath === null || $nodeUriPath === $targetNodeUriPath) {
+            // The page node path has not been changed
+            return;
+        }
+
+
+        $this->flushRoutingCacheForNode($targetNode);
+        $statusCode = (integer)$this->defaultStatusCode['redirect'];
+        $this->redirectionStorage->addRedirection($targetNodeUriPath, $nodeUriPath, $statusCode, $hosts);
+        /** @var ContentContext $contentContext */
+
+
+        $q = new FlowQuery([$node]);
+        foreach ($q->children('[instanceof TYPO3.Neos:Document]') as $childrenNode) {
```

Comment:
this only works for immediate children right? not recursively for grand children.. using `->find` would do that..